{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### **Title:** Preparing genotype variants and covariates for QTL analysis\n",
    "##### **Author:** Marliette Matos\n",
    "##### **Date:** 04/25/2025\n",
    "##### **Description:** Using genotype files QCed by Sam Gathan and filtered for a MAF>5%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Enviroment variables\n",
    "VCF=\"/gchm/cd4_QTL_analysis/01_genotype_snps_covar/02_genotype_covariates/analysis/001_merging_all_chr_vcf_MAF5/CD4_all_chr_ashkenazi.407.AF5.Q5.BA.vcf.gz\"\n",
    "OUTDIR=\"/gchm/cd4_QTL_analysis/01_genotype_snps_covar/02_genotype_covariates/analysis/002.v4_calculating_pcs_MAF5/test2\"\n",
    "OUTLIERS_IID='/gchm/cd4_QTL_analysis/01_genotype_snps_covar/02_genotype_covariates/scripts/sam_MAF5/v4_removing_PC_outliers/CD4.allsamples.allchrs.snps.geno.mind.af1.hwe.kingcutoff.0884.king.cutoff.out.id'\n",
    "#these outliers samples are related samples plus 4 outliers\n",
    "INVERSION_SITES=\"/gchm/cd4_aging_genotypes/wgs_QC/scripts/inversion.txt\"\n",
    "COMMON_SAMPLES=\"/gchm/cd4_QTL_analysis/02_Gene_expression/analysis/001_preparing_seurat_obj_QTL/003_common_samples_gfgex_wgs.in.tsv\"\n",
    "#common samples between WGS and Gene Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert to vcf to plink format and removing related samples\n",
    "!plink2 --vcf $VCF \\\n",
    "--const-fid 0 \\\n",
    "--ref-allele $VCF 4 3 '#' \\\n",
    "--memory 120000 \\\n",
    "--remove $OUTLIERS_IID \\\n",
    "--make-bed --out $OUTDIR/CD4_all_chr_ashkenazi.384.AF5.Q5.BA.king \\\n",
    "--new-id-max-allele-len 198 \\\n",
    "--set-all-var-ids @:#[b38]\\$r,\\$a "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keeping only common samples  \n",
    "!plink2 --bfile $OUTDIR/CD4_all_chr_ashkenazi.384.AF5.Q5.BA.king \\\n",
    "--memory 12000 \\\n",
    "--keep $COMMON_SAMPLES \\\n",
    "--make-bed --out $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2 \\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prune for Hardy Weinberg Variants\n",
    "!plink2 --bfile $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2 \\\n",
    "--memory 12000 \\\n",
    "--hwe 1e-6 \\\n",
    "--make-bed --out $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LD prunning highly correlated variants to aid the computation\n",
    "!plink2 --bfile $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe \\\n",
    "--memory 12000 \\\n",
    "--exclude $INVERSION_SITES \\\n",
    "--indep-pairwise 50 5 0.2 \\\n",
    "--out $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe.indepSNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate PCs\n",
    "!plink2 --bfile $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe \\\n",
    "--extract $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe.indepSNP.prune.in \\\n",
    "--pca 50 \\\n",
    "--out $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe.ld"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genetic relatedness matrix (GRM)\n",
    "!plink2 --bfile $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe \\\n",
    "--extract $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe.indepSNP.prune.in \\\n",
    "--make-rel 'square' \\\n",
    "--out $OUTDIR/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe.grm "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before piping the resulting into tensorqTL, the variants should be renamed by chr/post/ref/alt -> 003_change_var_names.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Splittting the pre-pruned by chromosomes for cis-eQTL calling -> 004_splitting_genotyped_by_chr.sh"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r-reticulate",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
