{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1c44c78e-364d-443f-a5cb-ae295810e876",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54b0efb1-122a-49da-8f84-052c654b81e5",
   "metadata": {},
   "source": [
    "#### Filter the Plink variants for just SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbac023b-aaf3-4dfc-98dc-a9c3c479babb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: plink2: command not found\n"
     ]
    }
   ],
   "source": [
    "variants_plink=\"/gchm/cd4_QTL_analysis/01_genotype_snps_covar/02_genotype_covariates/analysis/002.v4_calculating_pcs_MAF5/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe.annot\"\n",
    "vars_out=\"input/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe.annot.snps\"\n",
    "\n",
    "!plink2 --bfile $variants_plink --snps-only just-acgt --make-bed --out ${vars_out}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b262a45d-00b3-44e5-8466-956c4deef2a3",
   "metadata": {},
   "source": [
    "#### Read the SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5d3cfe15-1827-4a00-b301-ec3fa1c351cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of          chr           variant_id  genetic_dist        pos allele1 allele2\n",
       "0         10     10:61172[b38]A,G             0      61172       G       A\n",
       "1         10     10:65673[b38]C,T             0      65673       T       C\n",
       "2         10     10:68110[b38]G,C             0      68110       C       G\n",
       "3         10     10:70493[b38]C,A             0      70493       A       C\n",
       "4         10     10:76277[b38]G,T             0      76277       T       G\n",
       "...      ...                  ...           ...        ...     ...     ...\n",
       "5810462    9  9:138201870[b38]C,A             0  138201870       A       C\n",
       "5810463    9  9:138202443[b38]C,T             0  138202443       T       C\n",
       "5810464    9  9:138202814[b38]C,T             0  138202814       T       C\n",
       "5810465    9  9:138206150[b38]G,A             0  138206150       A       G\n",
       "5810466    9  9:138208331[b38]T,G             0  138208331       G       T\n",
       "\n",
       "[5810467 rows x 6 columns]>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_table(\"input/CD4_all_chr_ashkenazi.364.AF1.QC.BA.king2.hwe.annot.snps.bim\",\n",
    "                         sep='\\t', header=None)\n",
    "df.columns = ['chr', 'variant_id', 'genetic_dist', 'pos', 'allele1', 'allele2']\n",
    "\n",
    "df.head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c05ca71d-98a2-486a-a21d-ff95e081b3f3",
   "metadata": {},
   "source": [
    "#### Edit the variant_ids and also set allele1/2 as Ref/Atl, not major/minor like Plink format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "065b3548-3331-402e-8485-72cbd41b348f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>variant_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>61172</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr10_61172_A_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>65673</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10_65673_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>68110</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr10_68110_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>70493</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr10_70493_C_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>76277</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10_76277_G_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5810462</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138201870</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr9_138201870_C_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5810463</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138202443</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr9_138202443_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5810464</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138202814</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr9_138202814_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5810465</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138206150</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr9_138206150_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5810466</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138208331</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr9_138208331_T_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5810467 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr        pos allele1 allele2          variant_id\n",
       "0        chr10      61172       A       G     chr10_61172_A_G\n",
       "1        chr10      65673       C       T     chr10_65673_C_T\n",
       "2        chr10      68110       G       C     chr10_68110_G_C\n",
       "3        chr10      70493       C       A     chr10_70493_C_A\n",
       "4        chr10      76277       G       T     chr10_76277_G_T\n",
       "...        ...        ...     ...     ...                 ...\n",
       "5810462   chr9  138201870       C       A  chr9_138201870_C_A\n",
       "5810463   chr9  138202443       C       T  chr9_138202443_C_T\n",
       "5810464   chr9  138202814       C       T  chr9_138202814_C_T\n",
       "5810465   chr9  138206150       G       A  chr9_138206150_G_A\n",
       "5810466   chr9  138208331       T       G  chr9_138208331_T_G\n",
       "\n",
       "[5810467 rows x 5 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract alleles from variant_id\n",
    "variants = df.copy()\n",
    "variants[['allele1', 'allele2']] = variants['variant_id'].str.extract(r'.*\\](.*),(.*)\\s*$')\n",
    "\n",
    "# Create new variant_id format with 'chr' prefix\n",
    "variants['variant_id'] = variants['variant_id'].apply(\n",
    "    lambda x: re.sub(r'(\\d+):(\\d+)\\[b38\\](.*),(.*)\\s*', r'chr\\1_\\2_\\3_\\4', x) if pd.notna(x) else x\n",
    ")\n",
    "\n",
    "# Select and reorder columns\n",
    "variants = variants[['chr', 'pos', 'allele1', 'allele2', 'variant_id']]\n",
    "variants['chr'] = 'chr' + variants['chr'].astype(str)\n",
    "variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ada0a34b-a216-4157-8996-afd9404d2af9",
   "metadata": {},
   "source": [
    "#### Read the peaks from chrombpnet and filter variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d6d149d6-9f1f-4429-9cc0-b55d6d05c866",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10003</td>\n",
       "      <td>10604</td>\n",
       "      <td>merged_cd4_samples_peak_1</td>\n",
       "      <td>3154</td>\n",
       "      <td>.</td>\n",
       "      <td>9.48084</td>\n",
       "      <td>315.49900</td>\n",
       "      <td>313.14900</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10996</td>\n",
       "      <td>11452</td>\n",
       "      <td>merged_cd4_samples_peak_2</td>\n",
       "      <td>399</td>\n",
       "      <td>.</td>\n",
       "      <td>3.08842</td>\n",
       "      <td>39.93770</td>\n",
       "      <td>38.00440</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15486</td>\n",
       "      <td>15745</td>\n",
       "      <td>merged_cd4_samples_peak_3</td>\n",
       "      <td>176</td>\n",
       "      <td>.</td>\n",
       "      <td>2.45185</td>\n",
       "      <td>17.60840</td>\n",
       "      <td>15.79440</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17140</td>\n",
       "      <td>17900</td>\n",
       "      <td>merged_cd4_samples_peak_4</td>\n",
       "      <td>300</td>\n",
       "      <td>.</td>\n",
       "      <td>3.00455</td>\n",
       "      <td>30.02770</td>\n",
       "      <td>28.13810</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>20304</td>\n",
       "      <td>21778</td>\n",
       "      <td>merged_cd4_samples_peak_5a</td>\n",
       "      <td>31</td>\n",
       "      <td>.</td>\n",
       "      <td>1.52875</td>\n",
       "      <td>3.19981</td>\n",
       "      <td>1.61032</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125949</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50702244</td>\n",
       "      <td>50702805</td>\n",
       "      <td>merged_cd4_samples_peak_61127</td>\n",
       "      <td>249</td>\n",
       "      <td>.</td>\n",
       "      <td>3.04065</td>\n",
       "      <td>24.96390</td>\n",
       "      <td>23.10080</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125950</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50731899</td>\n",
       "      <td>50732864</td>\n",
       "      <td>merged_cd4_samples_peak_61128</td>\n",
       "      <td>4998</td>\n",
       "      <td>.</td>\n",
       "      <td>11.15230</td>\n",
       "      <td>499.81700</td>\n",
       "      <td>497.33900</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125951</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50756397</td>\n",
       "      <td>50757597</td>\n",
       "      <td>merged_cd4_samples_peak_61129</td>\n",
       "      <td>33759</td>\n",
       "      <td>.</td>\n",
       "      <td>20.33370</td>\n",
       "      <td>3375.95000</td>\n",
       "      <td>3372.14000</td>\n",
       "      <td>648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125952</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50775188</td>\n",
       "      <td>50775523</td>\n",
       "      <td>merged_cd4_samples_peak_61130</td>\n",
       "      <td>76</td>\n",
       "      <td>.</td>\n",
       "      <td>1.96748</td>\n",
       "      <td>7.67437</td>\n",
       "      <td>5.96979</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125953</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50783044</td>\n",
       "      <td>50784296</td>\n",
       "      <td>merged_cd4_samples_peak_61131</td>\n",
       "      <td>30880</td>\n",
       "      <td>.</td>\n",
       "      <td>14.51720</td>\n",
       "      <td>3088.06000</td>\n",
       "      <td>3084.42000</td>\n",
       "      <td>657</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>125954 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0         1         2                              3      4  5  \\\n",
       "0        chr1     10003     10604      merged_cd4_samples_peak_1   3154  .   \n",
       "1        chr1     10996     11452      merged_cd4_samples_peak_2    399  .   \n",
       "2        chr1     15486     15745      merged_cd4_samples_peak_3    176  .   \n",
       "3        chr1     17140     17900      merged_cd4_samples_peak_4    300  .   \n",
       "4        chr1     20304     21778     merged_cd4_samples_peak_5a     31  .   \n",
       "...       ...       ...       ...                            ...    ... ..   \n",
       "125949  chr22  50702244  50702805  merged_cd4_samples_peak_61127    249  .   \n",
       "125950  chr22  50731899  50732864  merged_cd4_samples_peak_61128   4998  .   \n",
       "125951  chr22  50756397  50757597  merged_cd4_samples_peak_61129  33759  .   \n",
       "125952  chr22  50775188  50775523  merged_cd4_samples_peak_61130     76  .   \n",
       "125953  chr22  50783044  50784296  merged_cd4_samples_peak_61131  30880  .   \n",
       "\n",
       "               6           7           8    9  \n",
       "0        9.48084   315.49900   313.14900  110  \n",
       "1        3.08842    39.93770    38.00440  282  \n",
       "2        2.45185    17.60840    15.79440  186  \n",
       "3        3.00455    30.02770    28.13810  334  \n",
       "4        1.52875     3.19981     1.61032   70  \n",
       "...          ...         ...         ...  ...  \n",
       "125949   3.04065    24.96390    23.10080  186  \n",
       "125950  11.15230   499.81700   497.33900  439  \n",
       "125951  20.33370  3375.95000  3372.14000  648  \n",
       "125952   1.96748     7.67437     5.96979  171  \n",
       "125953  14.51720  3088.06000  3084.42000  657  \n",
       "\n",
       "[125954 rows x 10 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks=pd.read_table(\"~/cd4_chrombpnet/data/inputs/peaks/merged_cd4_samples_peaks_no_blacklist.sorted.bed\",\n",
    "                         sep='\\t', header=None)\n",
    "peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "cab2d5e6-6f77-4817-ad4f-5ddcc9b620e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "\n",
    "def filter_variants_within_peaks(variants_df, peaks_df):\n",
    "    # Create temporary copies with renamed columns for PyRanges\n",
    "    variants_pr_df = variants_df.copy()\n",
    "    peaks_pr_df = peaks_df.copy()\n",
    "    \n",
    "    # Rename for PyRanges (but don't modify the original)\n",
    "    variants_pr_df = variants_pr_df.rename(columns={'chr': 'Chromosome', 'pos': 'Start'})\n",
    "    variants_pr_df['End'] = variants_pr_df['Start'] - 1 + 1  # Convert position to 0-indexed and then add 1 for end\n",
    "    \n",
    "    peaks_pr_df = peaks_pr_df.rename(columns={\n",
    "        '0': 'Chromosome',\n",
    "        '1': 'Start',\n",
    "        '2': 'End'\n",
    "    })\n",
    "    \n",
    "    # Create PyRanges objects\n",
    "    variants_pr = pr.PyRanges(variants_pr_df)\n",
    "    peaks_pr = pr.PyRanges(peaks_pr_df)\n",
    "    \n",
    "    # Find overlaps\n",
    "    overlaps = variants_pr.join(peaks_pr)\n",
    "    \n",
    "    # Extract just the variant_ids\n",
    "    variant_ids_in_peaks = set(overlaps.df['variant_id'])\n",
    "    \n",
    "    # Filter the ORIGINAL variants dataframe to maintain all original columns\n",
    "    filtered_variants = variants_df[variants_df['variant_id'].isin(variant_ids_in_peaks)]\n",
    "    \n",
    "    return filtered_variants\n",
    "\n",
    "filtered_variants = find_variants_in_peaks(variants, peaks)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5d139ff2-5f31-425c-81a4-5a92ee0201ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>original_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17558</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17559_G_C</td>\n",
       "      <td>17559</td>\n",
       "      <td>17140</td>\n",
       "      <td>17900</td>\n",
       "      <td>merged_cd4_samples_peak_4</td>\n",
       "      <td>300</td>\n",
       "      <td>.</td>\n",
       "      <td>3.00455</td>\n",
       "      <td>30.02770</td>\n",
       "      <td>28.13810</td>\n",
       "      <td>334</td>\n",
       "      <td>17559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183802</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_183803_G_A</td>\n",
       "      <td>183803</td>\n",
       "      <td>183260</td>\n",
       "      <td>184667</td>\n",
       "      <td>merged_cd4_samples_peak_17a</td>\n",
       "      <td>35</td>\n",
       "      <td>.</td>\n",
       "      <td>1.42319</td>\n",
       "      <td>3.51459</td>\n",
       "      <td>1.91425</td>\n",
       "      <td>93</td>\n",
       "      <td>183803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183802</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_183803_G_A</td>\n",
       "      <td>183803</td>\n",
       "      <td>183260</td>\n",
       "      <td>184667</td>\n",
       "      <td>merged_cd4_samples_peak_17b</td>\n",
       "      <td>219</td>\n",
       "      <td>.</td>\n",
       "      <td>2.33050</td>\n",
       "      <td>21.93830</td>\n",
       "      <td>20.09420</td>\n",
       "      <td>573</td>\n",
       "      <td>183803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183802</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_183803_G_A</td>\n",
       "      <td>183803</td>\n",
       "      <td>183260</td>\n",
       "      <td>184667</td>\n",
       "      <td>merged_cd4_samples_peak_17c</td>\n",
       "      <td>193</td>\n",
       "      <td>.</td>\n",
       "      <td>2.21077</td>\n",
       "      <td>19.39190</td>\n",
       "      <td>17.56460</td>\n",
       "      <td>969</td>\n",
       "      <td>183803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183802</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_183803_G_A</td>\n",
       "      <td>183803</td>\n",
       "      <td>183260</td>\n",
       "      <td>184667</td>\n",
       "      <td>merged_cd4_samples_peak_17d</td>\n",
       "      <td>170</td>\n",
       "      <td>.</td>\n",
       "      <td>2.11368</td>\n",
       "      <td>17.01890</td>\n",
       "      <td>15.21020</td>\n",
       "      <td>1133</td>\n",
       "      <td>183803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242039</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50674805</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr22_50674806_G_C</td>\n",
       "      <td>50674806</td>\n",
       "      <td>50674661</td>\n",
       "      <td>50675786</td>\n",
       "      <td>merged_cd4_samples_peak_61125</td>\n",
       "      <td>795</td>\n",
       "      <td>.</td>\n",
       "      <td>3.79752</td>\n",
       "      <td>79.56260</td>\n",
       "      <td>77.51180</td>\n",
       "      <td>430</td>\n",
       "      <td>50674806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242040</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50775330</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr22_50775331_G_C</td>\n",
       "      <td>50775331</td>\n",
       "      <td>50775188</td>\n",
       "      <td>50775523</td>\n",
       "      <td>merged_cd4_samples_peak_61130</td>\n",
       "      <td>76</td>\n",
       "      <td>.</td>\n",
       "      <td>1.96748</td>\n",
       "      <td>7.67437</td>\n",
       "      <td>5.96979</td>\n",
       "      <td>171</td>\n",
       "      <td>50775331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242041</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50775397</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr22_50775398_G_A</td>\n",
       "      <td>50775398</td>\n",
       "      <td>50775188</td>\n",
       "      <td>50775523</td>\n",
       "      <td>merged_cd4_samples_peak_61130</td>\n",
       "      <td>76</td>\n",
       "      <td>.</td>\n",
       "      <td>1.96748</td>\n",
       "      <td>7.67437</td>\n",
       "      <td>5.96979</td>\n",
       "      <td>171</td>\n",
       "      <td>50775398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242042</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50783302</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr22_50783303_T_C</td>\n",
       "      <td>50783303</td>\n",
       "      <td>50783044</td>\n",
       "      <td>50784296</td>\n",
       "      <td>merged_cd4_samples_peak_61131</td>\n",
       "      <td>30880</td>\n",
       "      <td>.</td>\n",
       "      <td>14.51720</td>\n",
       "      <td>3088.06000</td>\n",
       "      <td>3084.42000</td>\n",
       "      <td>657</td>\n",
       "      <td>50783303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242043</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50783671</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22_50783672_G_T</td>\n",
       "      <td>50783672</td>\n",
       "      <td>50783044</td>\n",
       "      <td>50784296</td>\n",
       "      <td>merged_cd4_samples_peak_61131</td>\n",
       "      <td>30880</td>\n",
       "      <td>.</td>\n",
       "      <td>14.51720</td>\n",
       "      <td>3088.06000</td>\n",
       "      <td>3084.42000</td>\n",
       "      <td>657</td>\n",
       "      <td>50783672</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>242044 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start allele1 allele2          variant_id       End  \\\n",
       "0            chr1     17558       G       C      chr1_17559_G_C     17559   \n",
       "1            chr1    183802       G       A     chr1_183803_G_A    183803   \n",
       "2            chr1    183802       G       A     chr1_183803_G_A    183803   \n",
       "3            chr1    183802       G       A     chr1_183803_G_A    183803   \n",
       "4            chr1    183802       G       A     chr1_183803_G_A    183803   \n",
       "...           ...       ...     ...     ...                 ...       ...   \n",
       "242039      chr22  50674805       G       C  chr22_50674806_G_C  50674806   \n",
       "242040      chr22  50775330       G       C  chr22_50775331_G_C  50775331   \n",
       "242041      chr22  50775397       G       A  chr22_50775398_G_A  50775398   \n",
       "242042      chr22  50783302       T       C  chr22_50783303_T_C  50783303   \n",
       "242043      chr22  50783671       G       T  chr22_50783672_G_T  50783672   \n",
       "\n",
       "         Start_b     End_b                              3      4  5         6  \\\n",
       "0          17140     17900      merged_cd4_samples_peak_4    300  .   3.00455   \n",
       "1         183260    184667    merged_cd4_samples_peak_17a     35  .   1.42319   \n",
       "2         183260    184667    merged_cd4_samples_peak_17b    219  .   2.33050   \n",
       "3         183260    184667    merged_cd4_samples_peak_17c    193  .   2.21077   \n",
       "4         183260    184667    merged_cd4_samples_peak_17d    170  .   2.11368   \n",
       "...          ...       ...                            ...    ... ..       ...   \n",
       "242039  50674661  50675786  merged_cd4_samples_peak_61125    795  .   3.79752   \n",
       "242040  50775188  50775523  merged_cd4_samples_peak_61130     76  .   1.96748   \n",
       "242041  50775188  50775523  merged_cd4_samples_peak_61130     76  .   1.96748   \n",
       "242042  50783044  50784296  merged_cd4_samples_peak_61131  30880  .  14.51720   \n",
       "242043  50783044  50784296  merged_cd4_samples_peak_61131  30880  .  14.51720   \n",
       "\n",
       "                 7           8     9  original_pos  \n",
       "0         30.02770    28.13810   334         17559  \n",
       "1          3.51459     1.91425    93        183803  \n",
       "2         21.93830    20.09420   573        183803  \n",
       "3         19.39190    17.56460   969        183803  \n",
       "4         17.01890    15.21020  1133        183803  \n",
       "...            ...         ...   ...           ...  \n",
       "242039    79.56260    77.51180   430      50674806  \n",
       "242040     7.67437     5.96979   171      50775331  \n",
       "242041     7.67437     5.96979   171      50775398  \n",
       "242042  3088.06000  3084.42000   657      50783303  \n",
       "242043  3088.06000  3084.42000   657      50783672  \n",
       "\n",
       "[242044 rows x 16 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "44ba2ad8-4d15-4199-9477-fc16c57e1a69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean variants per peak: 2.91\n",
      "Median variants per peak: 2.00\n",
      "Maximum variants in a peak: 50\n",
      "Minimum variants in a peak: 1\n",
      "Total peaks containing variants: 83049\n"
     ]
    }
   ],
   "source": [
    "filtered_variants.shape\n",
    "filtered_variants['variant_id'].nunique()\n",
    "filtered_variants[3].nunique()\n",
    "\n",
    "def calculate_mean_variants_per_peak(filtered_variants_df):\n",
    "    # Count variants per peak\n",
    "    variants_per_peak = filtered_variants_df.groupby(3).size()\n",
    "    \n",
    "    # Calculate mean\n",
    "    mean_variants = variants_per_peak.mean()\n",
    "    \n",
    "    # Calculate additional statistics if needed\n",
    "    median_variants = variants_per_peak.median()\n",
    "    max_variants = variants_per_peak.max()\n",
    "    min_variants = variants_per_peak.min()\n",
    "    \n",
    "    return {\n",
    "        'mean': mean_variants,\n",
    "        'median': median_variants,\n",
    "        'max': max_variants,\n",
    "        'min': min_variants,\n",
    "        'total_peaks_with_variants': len(variants_per_peak)\n",
    "    }\n",
    "# Calculate mean variants per peak\n",
    "stats = calculate_mean_variants_per_peak(filtered_variants)\n",
    "\n",
    "print(f\"Mean variants per peak: {stats['mean']:.2f}\")\n",
    "print(f\"Median variants per peak: {stats['median']:.2f}\")\n",
    "print(f\"Maximum variants in a peak: {stats['max']}\")\n",
    "print(f\"Minimum variants in a peak: {stats['min']}\")\n",
    "print(f\"Total peaks containing variants: {stats['total_peaks_with_variants']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3ff988a4-74d4-4d8c-972c-3716caf65b49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(160248, 5)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of            chr        pos allele1 allele2          variant_id\n",
       "4        chr10      76277       G       T     chr10_76277_G_T\n",
       "9        chr10      87333       G       C     chr10_87333_G_C\n",
       "10       chr10      87401       C       T     chr10_87401_C_T\n",
       "11       chr10      87928       G       A     chr10_87928_G_A\n",
       "12       chr10      87983       T       C     chr10_87983_T_C\n",
       "...        ...        ...     ...     ...                 ...\n",
       "5810266   chr9  138022809       G       T  chr9_138022809_G_T\n",
       "5810267   chr9  138022939       G       C  chr9_138022939_G_C\n",
       "5810392   chr9  138127843       A       G  chr9_138127843_A_G\n",
       "5810393   chr9  138128413       C       A  chr9_138128413_C_A\n",
       "5810427   chr9  138147679       G       C  chr9_138147679_G_C\n",
       "\n",
       "[160248 rows x 5 columns]>"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ids_in_peaks = set(filtered_variants['variant_id'])\n",
    "# Subset the original variants table using these IDs\n",
    "variants_in_peaks = variants[variants['variant_id'].isin(variant_ids_in_peaks)]\n",
    "print(variants_in_peaks.shape)\n",
    "variants_in_peaks.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9c3e53f3-604f-429a-934e-8feb046d4689",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to TSV file\n",
    "variants_in_peaks.to_csv('input/converted_variants_inpeaks.tsv', sep='\\t', index=False, header=None)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r-reticulate",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
